File: src/javascript/algorithms/matrix/maximalSquare/maximalSquare.js
/**
 * Maximal Square (LeetCode 221)
 *
 * Find the largest square of 1s in a binary matrix and return its area.
 *
 * Approach: Dynamic Programming
 * TC: O(m*n)  SC: O(m*n)
 */

function maximalSquare(matrix) {
    const rows = matrix.length, cols = matrix[0].length;
    const dp = Array.from({ length: rows }, () => new Array(cols).fill(0));
    let maxSide = 0;

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (matrix[r][c] === '1') {
                if (r === 0 || c === 0) {
                    dp[r][c] = 1;
                } else {
                    dp[r][c] = 1 + Math.min(dp[r-1][c], dp[r][c-1], dp[r-1][c-1]);
                }
                maxSide = Math.max(maxSide, dp[r][c]);
            }
        }
    }

    return maxSide * maxSide;
}

const testCases = [
    {
        matrix: [['1','0','1','0','0'],['1','0','1','1','1'],['1','1','1','1','1'],['1','0','0','1','0']],
        expected: 4, label: "2x2 square"
    },
    { matrix: [['0','1'],['1','0']], expected: 1, label: "diagonal 1s" },
    { matrix: [['0']],              expected: 0, label: "all zeros" },
    { matrix: [['1','1','1'],['1','1','1'],['1','1','1']], expected: 9, label: "all ones 3x3" },
    { matrix: [['1']],              expected: 1, label: "single 1" },
];

testCases.forEach(({ matrix, expected, label }) => {
    const result = maximalSquare(matrix);
    const status = result === expected ? "PASS" : "FAIL";
    console.log(`  [${status}] ${label}: ${result} (expected ${expected})`);
});

File: src/python/algorithms/matrix/maximalSquare/maximal_square.py
"""
Maximal Square (LeetCode 221)

Find the largest square of 1s in a binary matrix and return its area.

Approach: Dynamic Programming
TC: O(m*n)  SC: O(m*n)
"""


def maximal_square(matrix: list[list[str]]) -> int:
    rows, cols = len(matrix), len(matrix[0])
    dp = [[0] * cols for _ in range(rows)]
    max_side = 0

    for r in range(rows):
        for c in range(cols):
            if matrix[r][c] == '1':
                if r == 0 or c == 0:
                    dp[r][c] = 1
                else:
                    dp[r][c] = 1 + min(dp[r-1][c], dp[r][c-1], dp[r-1][c-1])
                max_side = max(max_side, dp[r][c])

    return max_side * max_side


if __name__ == "__main__":
    test_cases = [
        (
            [['1','0','1','0','0'],
             ['1','0','1','1','1'],
             ['1','1','1','1','1'],
             ['1','0','0','1','0']],
            4, "2x2 square"
        ),
        ([['0','1'],['1','0']], 1, "diagonal 1s"),
        ([['0']],               0, "all zeros"),
        ([['1','1','1'],['1','1','1'],['1','1','1']], 9, "all ones 3x3"),
        ([['1']],               1, "single 1"),
    ]

    for matrix, expected, label in test_cases:
        result = maximal_square(matrix)
        status = "PASS" if result == expected else "FAIL"
        print(f"  [{status}] {label}: {result} (expected {expected})")
